[FLINK-11855] Fix race condition in EmbeddedLeaderService#GrantLeadershipCall

tillrohrmann opened the pull request:
## What is the purpose of the change

Fix the race condition between executing EmbeddedLeaderService#GrantLeadershipCall
and a concurrent shutdown of the leader service by making GrantLeadershipCall not
accessing mutable state outside of a lock.

## Verifying this change

- Added `EmbeddedLeaderServiceTest#testConcurrentGrantLeadershipAndShutdown`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedLeaderService.java:
```diff
@@ -506,33 +515,28 @@ public void run() {
 			catch (Throwable t) {
 				logger.warn("Error granting leadership to contender", t);
 				contender.handleError(t instanceof Exception ? (Exception) t : new Exception(t));
-				leaderElectionService.isLeader = false;
```
The flag is no longer reset if the grant fails.

I'm wondering whether we need this flag at all. It's only being used in `EmbeddedLeaderElectionService#hasLeadership`, but shouldn't the `leaderSessionId` be sufficient for checking that? This seems especially true now that `updateLeader` updates both `isLeader` and `currentLeaderSessionId` right after another.

tillrohrmann replied:
I kept the flag because if we would remove it, then a user of the `EmbeddedLeaderElectionService#hasLeadership` could get a false positive if it guessed the current leadership id by chance.

It is true that we don't reset the `isLeader` flat to false in case of a failure. The assumption is that the `leaderContender` will stop the leader election service in case of a failure.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedLeaderServiceTest.java:
```diff
@@ -0,0 +1,67 @@
+			leaderElectionService.start(contender);
+			leaderElectionService.stop();
+
+			if (executor.numQueuedRunnables() > 0) {
```
There are no safeguards since it is actually an implementation detail how the `LeaderContender` is granted leadership. Also that `start` will directly send a `GrantLeadershipCall` is an implementation detail on which we rely by not having this check. Strictly speaking also this check does not guarantee that if the `ManuallyTriggeredScheduledExecutor` has a runnable enqueued that this is the `GrantLeadershipCall` runnable. Therefore, I opted for this compromise.

tillrohrmann replied:
But I think you are right that we could do this check a little bit better.

tillrohrmann commented:
Thanks for the review @zentol. I think you are right that we should also fix the `RevokeLeadershipCall`. I've added a fixup and addressed your comments.

tillrohrmann commented:
Thanks for the review @zentol. Merging.
